blob/blob.go

```diff
@@ -326,6 +356,14 @@ type WriterOptions struct {
 	// Keys may not be empty, and are lowercased before being written.
 	// Duplicate case-insensitive keys (e.g., "foo" and "FOO") are an error.
 	Metadata map[string]string
+
+	// Callback is a callback that will be called at most once, before any data
+	// is written. asFunc can be used as an escape hatch for accessing
+	// provider-specific types.
+	// See the provider documentation for which type(s) are supported, and
+	// https://github.com/google/go-cloud/blob/master/internal/docs/design.md#escape-hatches
+	// for more details on escape hatches.
+	Callback func(asFunc func(interface{}) bool) error
```

Comment:
Naming: I would call this `OnStart` or some such.

(Same advice about linking to the design doc.)

Reply:
How about "BeforeWrite"?